### PR TITLE
[CSI driver] Implementation of staged mount for vhost block and FS volumes in VM mode

### PR DESCRIPTION
**Pull request by aikuchin**
Previously endpoints were started on NodePublish phase, this breaks hotplug of multiple disks to the same VM (causes recreation of endpoints when helpler pod chandges, and ends up in deadlock attemptiong to create new endpoint before the old wan was stopped).

This commit splits volume mount between NodeStage and NodePublish phases. So endpoint is started in NodeStage, and NodePublish just creates mount to requested pod. With this split change of hotplug pod doesn't cause endpoint stop (because it doesn't trigger NodeUnpublish) but intead results in remounts, avoiding deadlock and reconnects.

What is unclear:
- [x] ~should stage mount be implemented for other types (FS, nbd?)~
   Decided that only vhost endpoints (NBS and NFS) should use new scheme
- [x] ~how to get podId in NodeStage phase (from volume attributes?)~
   We don't use podId at all, use instead instanceId from volume attributes, if it is not provided then use old scheme
- [x] ~how to make tranfer from old scheme seamless~
   Volumes that have instanceId attrubute will use new scheme, others will continue using old scheme 
- [x] ~should we use separate stage directory or just use s.socketsDir~
   Grouped volumes in instanceId directories

### CODE REVIEW

File: cloud/blockstore/tools/csi_driver/internal/driver/node.go [hunk 1]
```diff
@@ -151,6 +154,39 @@ func (s *nodeService) NodeStageVolume(
 			"VolumeCapability is missing in NodeStageVolumeRequest")
 	}
 
+	if s.vmMode {
+		nfsBackend := (req.VolumeContext[backendVolumeContextKey] == "nfs")
+
+		var err error
+		if instanceID := req.VolumeContext[instanceIDKey]; instanceID != "" {
```
Comment (antonmyagkov) [inline cloud/blockstore/tools/csi_driver/internal/driver/node.go:161]: Мне все-таки кажется что лучше использовать nodeId вместе instanceId, тогда не придется сохранять метаданные.

Можно просто логировать instanceId для дебага(возможно уже логируется вместе со всем VolumeContext)

Comment (antonmyagkov) [reply cloud/blockstore/tools/csi_driver/internal/driver/node.go:161]: решили оставить instanceId

File: cloud/blockstore/tools/csi_driver/internal/driver/node.go [hunk 2]
```diff
@@ -489,21 +706,32 @@ func (s *nodeService) nodeUnpublishVolume(
...
+	// In VM-mode for volumes that were staged we need just the unmount, so
+	// we could return here. Unfortunately we don't have enough information
+	// to know if this is a staged volume or a legacy one.
+	// Next StopEndpoint calls have no effect for staged volumes because their
+	// endpoints were started in socketsDir/instanceId/volumeId instead of
+	// socketsDir/podId/volumeId.
+	//
+	// When all VM disks are migrated to staged volumes we can enforce non-empty
+	// instanceId for new VM disks and add early return here:
+	// if s.vmMode { return nil }
```
[the remark this answers is not in the document]
Comment (antonmyagkov) [reply cloud/blockstore/tools/csi_driver/internal/driver/node.go:718]: Протестировал, unpublish ошибки не возвращает

File: cloud/blockstore/tools/csi_driver/internal/driver/node_test.go [hunk 3]
```diff
@@ -185,6 +190,193 @@ func TestPublishUnpublishFilestoreForKubevirt(t *testing.T) {
...
+		SocketPath: filepath.Join(socketsDir, podID, diskID, nfsSocketName),
+	}).Return(&nfs.TStopEndpointResponse{}, nil)
+
+	_, err = nodeService.NodeUnpublishVolume(ctx, &csi.NodeUnpublishVolumeRequest{
+		VolumeId:   diskID,
+		TargetPath: targetPath,
+	})
+	require.NoError(t, err)
+
+	// TODO: enable check that target directory is deleted
```
Comment (antonmyagkov) [inline cloud/blockstore/tools/csi_driver/internal/driver/node_test.go:338]: Добавь сюда номер тикета который ты создал, что бы мы не забыли включить эту проверку потом

Comment (aikuchin) [reply cloud/blockstore/tools/csi_driver/internal/driver/node_test.go]: added link to https://github.com/ydb-platform/nbs/issues/2120

**Review by antonmyagkov — APPROVED** [review]

File: cloud/blockstore/tools/csi_driver/internal/driver/node.go [hunk 4]
```diff
@@ -151,6 +154,39 @@ func (s *nodeService) NodeStageVolume(
...
+			if nfsBackend {
+				backend = "nfs"
+			}
+			if err = s.writeStageData(stageRecordPath, &StageData{
+				Backend:       backend,
+				InstanceId:    instanceID,
+				RealStagePath: s.getEndpointDir(instanceID, req.VolumeId),
+			}); err != nil {
+				return nil, s.statusErrorf(codes.Internal,
+					"Failed to wriete stage record: %v", err)
```
[the remark this answers is not in the document]
Comment (aikuchin) [reply cloud/blockstore/tools/csi_driver/internal/driver/node.go]: Fixed

**Review by drbasic — APPROVED** [review]

**Pull request merged**